### PR TITLE
Lukasz logupload

### PR DESCRIPTION
**Pull request by lchodar**
Refactoring of plugin code and added functionality to send all build logs to Sumo Logic.

### CODE REVIEW

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+import org.kohsuke.stapler.DataBoundConstructor;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.logging.Logger;
+
+/**
+ * Created by deven on 7/6/15.
+ *
+ * This publisher will sendLogs build metadata to a Sumologic HTTP collector.
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java]: We put this command usually first, before author and date.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+
+  @Override
+  public boolean perform(AbstractBuild<?, ?> build, Launcher launcher, BuildListener listener) throws InterruptedException, IOException {
+    send(build, listener);
+    return true;
+  }
+
+  @Override
+  public PluginDescriptorImpl getDescriptor() {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java]: No blank line here.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    return (PluginDescriptorImpl) super.getDescriptor();
+  }
+
+  @Override
+  public BuildStepMonitor getRequiredMonitorService() {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java]: No blank line here.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  protected void send(AbstractBuild build, TaskListener listener) {
+    Gson gson = new Gson();
+    String json = gson.toJson(ModelFactory.createBuildModel(build));
+    String projectName = build.getProject().getDisplayName();
+
+    LOG.info("Uploading build status to sumologic: " + json);
+    logSender.sendLogs(getDescriptor().getUrl(), json.getBytes(), projectName, "jenkinsStatus");
+  }
+
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumoBuildNotifier.java]: Would remove those empty lines near end.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.logging.Logger;
+
+/**
+ * Created by lukasz on 3/21/17.
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: Add one line comment what it does.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    logSender = LogSender.getInstance();
+
+    this.jobName = build.getProject().getDisplayName();
+    this.jobNumber = build.getDisplayName();
+    maxLinesPerBatch = descriptor.getMaxLinesInt();
+
+    currentLines = 0;
+    buffer = new ByteArrayBuffer(1);
+    this.url = descriptor.getUrl();
+  }
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: Add empty line below.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+  @Override
+  public void close() throws IOException {
+    super.close();
+    flushBuffer();
+  }
+
+  @Override
+  protected void eol(byte[] bytes, int i) throws IOException {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: Remove empty line.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+      return;
+    }
+
+    byte[] lines = buffer.toByteArray();
+    buffer.clear();
+    currentLines = 0;
+
+    try {
+      logSender.sendLogs(url, lines, jobName + jobNumber, "jenkinsStatus");
+    }
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: `} catch (...) {` in one line

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/TimestampingOutputStream.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+  public TimestampingOutputStream(OutputStream stream) {
+    super();
+    wrappedStream = stream;
+  }
+
+
+  @Override
+  protected void eol(byte[] bytes, int i) throws IOException {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/TimestampingOutputStream.java]: Would remove line.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/TimestampingOutputStream.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import java.io.IOException;
+import java.io.OutputStream;
+import java.io.PrintStream;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.logging.Logger;
+
+/**
+ * Created by lukasz on 3/21/17.
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/TimestampingOutputStream.java]: Would add one-line comment what it does.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    if (currentLines <= 0) {
+      return;
+    }
+
+    byte[] lines = buffer.toByteArray();
+    buffer.clear();
+    currentLines = 0;
+
+    try {
+      logSender.sendLogs(url, lines, jobName + jobNumber, "jenkinsStatus");
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: Shouldn't it be `jobName + "#" + jobNumber` or job number has `#` Would add comment about it or special variable.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/LogListener.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+import hudson.model.AbstractBuild;
+
+import java.io.IOException;
+import java.io.OutputStream;
+
+/**
+ * This class intercepts console output stream for every jenkins build and decorates it with additional functionality.
+ * Depending on configuration it can add two functionalities:
+ * 1) Add consistent timestamps to all output lines.
+ * 2) Forward all output logs to Sumo Logic.
```
Comment (davidcarltonsumo) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/LogListener.java]: Jenkins already has a mechanism for putting timestamps in log lines, can we piggyback on top of that?  (Or, if we can't, can we set it up so the timestamps added here only get sent to Sumo, not to the disk?  Assuming, that is, that the decorator affects all uses of the logs.)

Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/LogListener.java]: We can't piggy back. Jenkins timestamps are optional, just have hours.

Currently the plugin is designed as you proposed. We always add timestamp in logs sent to Sumo, but it's optional everywhere else.

Comment (lchodar) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/LogListener.java]: In addition to what Jacek wrote, there are plugins/dependencies that could do the same for us but we either way had to intercept whole log stream so can do it with one pass and without requireing external dependencies on other plugins.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java [hunk 13]
```diff
@@ -0,0 +1,173 @@
...
+  }
+
+  @Override
+  public boolean isApplicable(Class<? extends AbstractProject> aClass) {
+    return true;
+  }
+
+  @Override
+  public String getDisplayName() {
+    return "Sumologic build logger";
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java]: Sumo Logic

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+
+    this.queryPortal = queryPortal;
+  }
+
+  public boolean isBuildLogEnabled() {
+    return buildLogEnabled;
+  }
+
+  public void setBuildLogEnabled(boolean value) {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java]: Remove blank line.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+    return value;
+  }
+
+  public String getQueryPortal() {
+
+    return queryPortal;
+  }
+
+  public void setQueryPortal (String queryPortal) {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java]: Remove blank line.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java [hunk 16]
```diff
@@ -0,0 +1,173 @@
...
+      value = Integer.parseInt(maxLines);
+    } catch (NumberFormatException e) {
+      value = MAX_LINES_DEFAULT;
+    }
+
+    return value;
+  }
+
+  public String getQueryPortal() {
+
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/PluginDescriptorImpl.java]: Remove blank line.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+      return;
+    }
+
+    byte[] lines = buffer.toByteArray();
+    buffer.clear();
+    currentLines = 0;
+
+    try {
+      // jobNumber is a build number with #, e.g. #42
+      logSender.sendLogs(url, lines, jobName + jobNumber, "jenkinsStatus");
```
Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: Would have a constant string `"jenkinsStatus"`

Comment (jakozaur) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/SumologicOutputStream.java]: Use configuration.

File: src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/LogListener.java (hunk 12 above)
Comment (davidcarltonsumo) [inline src/main/java/com/sumologic/jenkins/jenkinssumologicplugin/sender/LogListener.java]: It is not the case that Jenkins timestamps just have hours: they have minutes and seconds, e.g. https://jenkins.kumoroku.com/job/Master-Update-Jenkins/844/console.  No date as printed there, though I suspect it's in the sidecar data on disk.  (The timestamps are stored different from the raw log data.)

At any rate, as long as it only goes to Sumo, that's fine with me.  Personally, I wouldn't have them even optionally go elsewhere - this plugin should be focused on Sumo, not on providing other bits of functionality.

**Pull request merged**